### ISSUE TITLE
[IMPROVEMENT] Upgrade path improvement

### ISSUE THREAD

**Issue by shuo-wu** [longhorn/longhorn#3087]
**Is your feature request related to a problem? Please describe.**
In #2582 and #3068, longhorn manager pod launching will get crashed by the CR update conflicts that happen in the upgrade path. The direct cause is that some CRs will be blindly updated as long as there is a new longhorn manager pod up, regardless of the system being upgraded or the CRs already being up-to-date.

The hotfix I applied is to the above 2 issues avoids the update when there is nothing change in the CRs. Then the upgrade failure won’t be triggered during fresh installation or longhorn pod restart. But it can still block some manager pods during the upgrade, .e.g., upgrade from 1.1.x to 1.2.x.

To fully resolve this issue, I think we can introduce the following:
1. Longhorn version recognition. Longhorn can skip all old upgrade paths if somehow it knows the previous version and the upgrading version.
2. By checking the setting values, all other longhorn manager pods can give up the leadership election and skip the upgrade paths if the 1st leader manager pod succeeds to execute the upgrade path. This may speed up the longhorn manager launching when there are hundreds of worker nodes in a cluster.

**Describe the solution you'd like**
We can add a new read-only setting `Current Longhorn version`, with which Longhorn would recognize which version it is upgraded from before walking through the upgrade paths, so that Longhorn will skip all old upgrade paths. And after upgrade paths execution (by **the 1st manager pod that takes over the leadership**), the setting will be updated. Later, once there are other manager pods getting started, they can realize the current Longhorn version is up to date and there is no need to take care of any upgrade path. As a result, the upgrade path will be executed once by the 1st manager pod before any CR controller starts, and there are no other components fighting for CR update.

**Describe alternatives you've considered**
N/A

**Additional context**
- Notice that the upgrade path idempotency should not be break even if the enhancement is introduced.

**Comment by derekbit** [comment]
Leaving a not here.

For future upgrade refactoring, need to separate the two scenarios

(1) CRD version upgrade
(2) CRD version not upgraded, but longhorn version upgrade

The logic of current upgrade mixes them up, so it is still messy now.

**Comment by PhanLe1010** [comment]
we need to test upgrading Longhorn in big cluster with many volumes, ref https://github.com/longhorn/longhorn/issues/3368#issuecomment-987867876. The expected time is 30s per node

**Comment by shuo-wu** [comment]
Test summary:

Env:
20 worker nodes (2v core, 4G RAM); 200 running volumes on different nodes

Result:
| Case | Result |
| :---- | :----: |
| Restarting all longhorn manager pods for the old version | 17mins |
| Upgrading Longhorn via the old upgrade path (the versions before and after the upgrade that do not contain the PR) | 17mins |
| Restarting all longhorn manager pods for the new version | 1~2mins |
| Upgrading Longhorn via the new upgrade path (the versions before and after the upgrade that both contain the PR) | 3~4mins |

**Comment by shuo-wu** [comment]
The idea for this improvement is pretty straightforward. Do we still need any doc or LEP?

**Comment by innobead** [comment]
> The idea for this improvement is pretty straightforward. Do we still need any doc or LEP?

If that's the case, no need.

**Comment by shuo-wu** [comment]
BTW, we need to backport this improvement to v1.2.4 and v1.1.4, right? @innobead

**Comment by innobead** [comment]
> BTW, we need to backport this improvement to v1.2.4 and v1.1.4, right? @innobead

Yeah, please.

**Comment by kaxing** [comment]
Will be tested with scalability test setup once the rc1 image is out for v1.3.0.

**Comment by kaxing** [comment]
Validating in progress, cc @shuo-wu 

Testing from v1.2.2 to v1.3.0-dev.
Retest with 20 volumes attached to 20 simple workload and concurrent-automatic-engine-upgrade-per-node-limit set to 3.
By measuring the `status.numberAvailable` pod of `daemonset/longhorn-manager` right after longhorn.yaml deployed.
Sampling the average time out of 4 rounds of test: (container image download time is ignored as the cluster was reused for each round of test.)
Version | Time (sec.)
:- | :-
Fresh install → v1.2.2 | 123
V1.2.2 → v1.2.3 | 343
v1.2.3 → v1.2.4 | 44
v1.2.4 -> v1.3.0-dev | 59

<sup>* Each round of test will wait for all engine image gets upgraded.</sup>

---
<Details>
<Summary>Expand to see the timer function for the reference</Summary>

```bash
ds-timer() {
  local start=`date +%s`
  echo "Timer Starts at $start"
  local desired_count=$(kubectl get daemonset/longhorn-manager -n longhorn-system -o json | jq '.status.desiredNumberScheduled')
  update_current_count() { echo $(kubectl get daemonset/longhorn-manager -n longhorn-system -o json | jq '.status.numberAvailable'); }
  local current_count=$(update_current_count)
  while [[ $desired_count -ne $current_count ]]
  do
    current_count=$(update_current_count)
    echo "longhorn-manager ready pod count: $current_count/$desired_count" | ts '[%Y-%m-%d_%H:%M:%S]'
  done
  local end=`date +%s`
  local elapsed=$((end-start))
  echo "Timer Ends at: $end"
  echo "Total elapsed time in second: $elapsed "
}
```

</Details>

**Comment by shuo-wu** [comment]
When there are multiple volumes in the system, upgrading to v1.3.0 is still relatively fast, right? 
BTW, you can launch 100+ volumes instead and see the difference.

**Comment by kaxing** [comment]
I redo the test with 100pod-volumes(size in 11MB) and the result still shows v1.2.4 had better deployment speed of longhorn-manager pods.

Gradual upgrade:
install path | time
:-|:-
fresh install v1.2.3 | 146s
upgrade to v1.2.4* | 56s
upgrade from v1.2.4 to v1.3.0-preview1 | 97s

<sup>* Some engine image will stuck when upgrading to v1.2.4, but it will be resolved when upgrade to 1.3.0-preview1 image</sup>

Jump upgrade:
install path | time
:-|:-
fresh install v1.2.3 | 137s
upgrade to v1.3.0-preview1 | 100s

**Comment by kaxing** [comment]
Validation - **PASSED**

Validated with v1.2.4 image, upgrade time had been improved since v1.2.3: https://github.com/longhorn/longhorn/issues/3087#issuecomment-1116121960

Also had reference test with v1.3.0-preview1(see previous comment): https://github.com/longhorn/longhorn/issues/3087#issuecomment-1120730706

As the improvement is made in v1.2.4 branch, the launch time is expecting differ with v1.3.0.
